gcore/gdalsubdatasetinfo.cpp

```diff
@@ -0,0 +1,137 @@
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *                                                                         *
+ ***************************************************************************/
+#include "gdalsubdatasetinfo.h"
+#include "gdal_priv.h"
+
+/************************************************************************/
+/*                     Subdataset informational functions               */
+/************************************************************************/
+
+GDALSubdatasetInfoH CPL_STDCALL GDALGetSubdatasetInfo(const char *pszFileName)
+{
+    // Iterate all drivers
+    GDALDriverManager *poDM_ = GetGDALDriverManager();
+    const int nDriverCount = poDM_->GetDriverCount();
+    for (int iDriver = 0; iDriver < nDriverCount; ++iDriver)
+    {
+        GDALDriver *poDriver = poDM_->GetDriver(iDriver);
+        char **papszMD = GDALGetMetadata(poDriver, nullptr);
+        if (!CPLFetchBool(papszMD, GDAL_DMD_SUBDATASETS, false) ||
+            !poDriver->pfnGetSubdatasetInfoFunc)
+        {
+            continue;
+        }
+
+        GDALSubdatasetInfo *poGetSubdatasetInfo =
+            poDriver->pfnGetSubdatasetInfoFunc(pszFileName);
+
+        if (!poGetSubdatasetInfo)
+        {
+            continue;
+        }
+
+        return static_cast<GDALSubdatasetInfoH>(poGetSubdatasetInfo);
+    }
+    return nullptr;
+}
+
+/************************************************************************/
+/*                       GDALDestroySubdatasetInfo()                    */
+/************************************************************************/
+
+/**
+ * \brief Destroys subdataset info.
+ *
+ * This function is the same as the C++ method GDALSubdatasetInfo::~GDALSubdatasetInfo()
+ */
+void GDALDestroySubdatasetInfo(GDALSubdatasetInfoH hInfo)
+
+{
+    delete hInfo;
+}
+
+char *GDALSubdatasetInfoGetFileName(GDALSubdatasetInfoH hInfo)
+{
+    return CPLStrdup(hInfo->GetFileName().c_str());
+}
+
+char *GDALSubdatasetInfoGetSubdatasetName(GDALSubdatasetInfoH hInfo)
+{
+    return CPLStrdup(hInfo->GetSubdatasetName().c_str());
+}
+
+char *GDALSubdatasetInfoModifyFileName(GDALSubdatasetInfoH hInfo,
+                                       const char *pszNewFileName)
+{
+    return CPLStrdup(hInfo->ModifyFileName(pszNewFileName).c_str());
+}
+
+GDALSubdatasetInfo::GDALSubdatasetInfo(const std::string &fileName)
+    : m_fileName(fileName), m_baseComponent(), m_subdatasetComponent(),
+      m_driverPrefixComponent()
+{
+}
+
+std::string GDALSubdatasetInfo::GetFileName() const
+{
+    if (!m_initialized)
+    {
+        GDALSubdatasetInfo *this_ = const_cast<GDALSubdatasetInfo *>(this);
+        this_->parseFileName();
+        m_initialized = true;
+    }
+    return m_driverPrefixComponent + ":" + m_baseComponent;
```

Comment:
I would expect this to be just:
```suggestion
    return m_filename;
```
I would avoid assuming at this generic level that a ':' is used to join m_driverPrefixComponent and m_baseComponent
I guess if there is a need to return m_driverPrefixComponent + ":" + m_baseComponent that should be through a virtual method implemented in the driver

Reply:
Ok, my idea was to return the path to the parent dataset with the subdataset information stripped off, I thought it was useful but we can add it later if needed.